org.jacoco.core/src/org/jacoco/core/internal/runtime/ClassInjectionRuntime.java

```diff
@@ -0,0 +1,297 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2018 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.runtime;
+
+import java.lang.instrument.Instrumentation;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.jacoco.core.runtime.AbstractRuntime;
+import org.jacoco.core.runtime.IRuntime;
+import org.jacoco.core.runtime.RuntimeData;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+/**
+ * {@link IRuntime} that uses introduced in Java 9
+ * {@code java.lang.instrument.Instrumentation.redefineModule} and
+ * {@code java.lang.invoke.MethodHandles.Lookup.defineClass}.
+ */
+public class ClassInjectionRuntime extends AbstractRuntime {
+
+	private static final String FIELD_NAME = "data";
+
+	private static final String FIELD_TYPE = "Ljava/lang/Object;";
+
+	private final String injectedClassName;
+
+	private final MethodHandle dataFieldSetter;
+
+	/**
+	 * Creates new instance of runtime.
+	 *
+	 * @param instrumentation
+	 *            instrumentation interface
+	 * @return new runtime instance or <code>null</code> if unable to create
+	 */
+	public static IRuntime create(final Instrumentation instrumentation) {
+		try {
+			redefineJavaBaseModule(instrumentation);
+			return new ClassInjectionRuntime(Object.class);
+		} catch (final Exception e) {
+			return null;
+		}
+	}
+
+	ClassInjectionRuntime(final Class<?> cls) throws Exception {
+		this.injectedClassName = cls.getName().replace('.', '/') + "$JaCoCo";
```

Comment:
I wonder whether a random class name would be more robust e.g. in case of multiple JaCoCo agents?

Reply:
> case of multiple JaCoCo agents

@marchof in #781 we agreed that case of multiple agents is completely misleading and error-prone, so IMO not an argument for random class name. Other things being equal, I prefer predictable behavior over random.

Reply:
Ok!